workflow/controller/dag.go

```diff
@@ -320,25 +320,15 @@ func (woc *wfOperationCtx) executeDAGTask(dagCtx *dagContext, taskName string) {
 	nodeName := dagCtx.taskNodeName(taskName)
 	for _, depName := range task.Dependencies {
 		depNode := dagCtx.GetTaskNode(depName)
-		if depNode != nil {
-			if depNode.Completed() {
-				depTask := dagCtx.getTask(depName)
-				// Run the node's onExit node, if any. Only nodes that have dependencies will have their onExit nodes
-				// executed here. Leaf nodes will have their onExit nodes executed above
-				hasOnExitNode, onExitNode, err := woc.runOnExitNode(depTask.Name, depTask.OnExit, dagCtx.boundaryID, dagCtx.tmplCtx)
-				if hasOnExitNode && (onExitNode == nil || !onExitNode.Completed() || err != nil) {
-					// The onExit node is either not complete or has errored out, return.
-					return
-				}
-
-				if !depNode.Successful() && !depTask.ContinuesOn(depNode.Phase) {
-					dependenciesSuccessful = false
-				}
-				continue
```

Comment:
Currently, this `continue` means that `executeDAGTask` will not be called on a node's dependency that has been completed, so we currently run its finalizer logic here. This complicates things unnecessarily. Instead we allow all nodes to call `executeDAGTask` and be detected as completed there.